metrics-loader: create metrics-loader tool

herkolategan opened the pull request:
This PR contains several commits to add a new tool to the arsenal, the `metrics-loader`.
The main purposes of the tool is to monitor a Google Storage Bucket (path) for OpenMetrics format files [1] and stream those metrics to a Prometheus remote-write endpoint [2].

The tool is built up out of several components. Each component's goal is described in more detail in the commit message.
A brief overview of all the components:
- Reader: Decodes an OpenMetrics file into Prometheus metric structs.
- Sender: Sends `decoded` metrics to a Prometheus remote-write endpoint
- Processor: Process a given input (File) by utilising a `Reader` &`Sender`.
- Storage: A simple layer to interact with a storage system (GCS in this case)
- Watcher: Watches the Google Bucket for changes, and invokes the Processor to process any new files.

[1] https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?rgh-link-date=2024-09-10T10%3A57%3A25Z
[2] https://prometheus.io/docs/specs/remote_write_spec/#prometheus-remote-write-specification

Resolves: #130403
Informs: #106662

Epic: None
Release note: None

herkolategan committed:
metrics-loader: create metrics-loader tool

Create the metrics-loader tool package, command and help message. The tool will
serve as a way to get metrics from `roachtest` into Bigtable via. a Prometheus
remote-write endpoint.

Epic: None
Release note: None

herkolategan committed:
metrics-loader: implement HTTP sender

`httpSender` implements a `Sender` interface for sending metrics to a remote
storage. The main post body consists of a `prompb.WriteRequest` that contains an
array of time series metric points. Additionally, the content is compressed with
snappy.

Epic: None
Release note: None

herkolategan committed:
metrics-loader: implement OpenMetrics reader

`openMetricsReader` implements a `Reader` interface for reading OpenMetrics file
data from an `io.Reader` to produce `prompb.Timeseries` output entries. The
entries are sent to a callback function so that the consumer can immediately
start utilising the output (start sending the data in parallel).

Epic: None
Release note: None

herkolategan committed:
metrics-loader: implement Processor

Implements the Processor, which utilises a `Reader` and a `Sender`, and serves
as the pipe layer between these two interfaces. Given an `io.Reader` the
Processor will read all `prompb.TimeSeries` metrics into a channel, and start
sending the data through to the provided `Sender`. Any error from the `Reader`
or `Sender` will cancel further processing for the given input and return an
error to the caller.

In addition, the `Processor` can do basic manipulations on the metrics if
required, like modifying the labels. This is useful if the remote-write storage
wants to label the source of the metrics.

Epic: None
Release note: None

herkolategan committed:
metrics-loader: implement Storage

The `Storage` interface provides a way of interacting with a file storage medium
to monitor for OpenMetrics files that need to be streamed to a remote-write
endpoint. It implements a few basic operations to manage a file storage system,
and be able to move files from a processing folder to a completed folder once
the file has been successfully sent to the remote-write endpoint.

Additionally, a `GoogleStorage` implementation is provided in order to register
for required notifications on a bucket. This will allow the application to
monitor a bucket for new unprocessed files.

Epic: None
Release note: None

herkolategan committed:
metrics-loader: monitor bucket with GoogleWatcher

This change add a Google storage bucket watcher that will detect any new
files for a given bucket and path, and use a `Processor` to process the files.
The watcher uses a pubsub subscription, to the bucket, to receive events
detailing new OpenMetrics files that have been added to the specified bucket
path. It first checks if there are existing subscriptions, otherwise will create
new ones.

Any errors reported by the `Processor` will result in the file not being moved
to a completed folder, but the application will continue running.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130574)
<!-- Reviewable:end -->

herkolategan committed:
metrics-loader: configure and run

This adds the required configuration via flags for all components to be able to
start the watcher. In the `RunE` function we configure a `Sender` with a given
remote-write URL, and use the default `OpenMetricsReader`, with the `Processor`
to parse OpenMetrics files. A GoogleWatcher is configured to start watching a
bucket via a PubSub subscription.

The app will continue to run until interrupted or if a serious error occurs
within the `watcher` which prevents if from monitoring the bucket.

Epic: None
Release note: None

DarrylWong reviewed (commented):
Should we add this to dev build? 

I tried adding something like:

 `"metrics-loader":      "//pkg/cmd/metrics-loader:metrics-loader",`
 
 [to the mapping](https://github.com/cockroachdb/cockroach/blob/41c3aa4084245e319ad92128f594706ea5fdc77f/pkg/cmd/dev/build.go#L78) but not sure what else needs to be done to get it to work.

DarrylWong commented:
Should we also add this new path to `.github/CODEOWNERS` under test-eng?

DarrylWong commented on pkg/cmd/metrics-loader/watcher.go:
```diff
@@ -0,0 +1,153 @@
+	if err != nil {
+		return err
+	}
+	return w.googleStorage.Move(uri, w.toProcessedPath(uri))
```
If `Process` already sends the metrics to prom, whats the reason for keeping the file around as opposed to deleting it?

DarrylWong commented:
On second thought, I wonder if this should live in its own repo. I don't see why this needs to live in the monorepo besides convenience but perhaps that's good enough reason 🤷 

Overall this looks good on first pass, but I haven't gotten around to actually trying it out myself.

herkolategan commented on pkg/cmd/metrics-loader/reader.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func (r *openMetricsReader) Read(reader io.Reader, outputFn ReaderOutputFn) error {
+	batch := make([]byte, 0, readBufferSize+readBatchSize)
```
Good catch! That was based on something else originally, I'll amend!

herkolategan commented:
> On second thought, I wonder if this should live in its own repo.
Good question, since it doesn't really have any dependencies in the main repo I'm not opposed to moving it under a new repo within cockroachdb. I just want to move it out of my own GH so others can contribute.

herkolategan commented:
Moving this to a different repo.
